File: frontend/src/api/finance.ts
import { apiGet } from './client';
import type { AccountsOverview, MonarchTransaction, BudgetOverview, CashflowSummary, RecurringItem, NetWorthHistory, SpendingBreakdown, CashflowTrend } from '../types/finance';

export function fetchAccounts() {
  return apiGet<AccountsOverview>('/api/finance/accounts');
}

export function fetchTransactions(limit = 50, offset = 0, startDate?: string, endDate?: string) {
  const params = new URLSearchParams({ limit: String(limit), offset: String(offset) });
  if (startDate) params.set('start_date', startDate);
  if (endDate) params.set('end_date', endDate);
  return apiGet<MonarchTransaction[]>(`/api/finance/transactions?${params}`);
}

export function fetchBudgets(startDate?: string, endDate?: string) {
  const params = new URLSearchParams();
  if (startDate) params.set('start_date', startDate);
  if (endDate) params.set('end_date', endDate);
  const qs = params.toString();
  return apiGet<BudgetOverview>(`/api/finance/budgets${qs ? `?${qs}` : ''}`);
}

export function fetchCashflow(startDate?: string, endDate?: string) {
  const params = new URLSearchParams();
  if (startDate) params.set('start_date', startDate);
  if (endDate) params.set('end_date', endDate);
  const qs = params.toString();
  return apiGet<CashflowSummary>(`/api/finance/cashflow${qs ? `?${qs}` : ''}`);
}

export function fetchRecurring() {
  return apiGet<RecurringItem[]>('/api/finance/recurring');
}

export function fetchNetWorthHistory(months = 12) {
  return apiGet<NetWorthHistory>(`/api/finance/net-worth-history?months=${months}`);
}

export function fetchSpendingBreakdown(months = 2) {
  return apiGet<SpendingBreakdown>(`/api/finance/spending-breakdown?months=${months}`);
}

export function fetchCashflowTrend(months = 6) {
  return apiGet<CashflowTrend>(`/api/finance/cashflow-trend?months=${months}`);
}

File: backend/api/finance.py
"""Personal finance endpoints — Monarch Money integration."""
from __future__ import annotations

from flask import Blueprint, request, jsonify

finance_bp = Blueprint("finance", __name__)


def _get_service():
    """Lazy import to avoid startup failure if creds aren't set."""
    from backend.services.monarch_service import (
        get_accounts, get_transactions, get_budgets, get_cashflow, get_recurring,
    )
    return get_accounts, get_transactions, get_budgets, get_cashflow, get_recurring


@finance_bp.route("/api/finance/accounts")
def accounts():
    try:
        get_accounts, *_ = _get_service()
        data = get_accounts()
        return jsonify(data)
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@finance_bp.route("/api/finance/transactions")
def transactions():
    try:
        _, get_transactions, *_ = _get_service()
        limit = request.args.get("limit", 50, type=int)
        offset = request.args.get("offset", 0, type=int)
        start_date = request.args.get("start_date")
        end_date = request.args.get("end_date")
        data = get_transactions(limit=limit, offset=offset,
                                start_date=start_date, end_date=end_date)
        return jsonify(data)
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@finance_bp.route("/api/finance/budgets")
def budgets():
    try:
        *_, get_budgets, _, _ = _get_service()
        start_date = request.args.get("start_date")
        end_date = request.args.get("end_date")
        data = get_budgets(start_date=start_date, end_date=end_date)
        return jsonify(data)
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@finance_bp.route("/api/finance/cashflow")
def cashflow():
    try:
        *_, get_cashflow, _ = _get_service()
        start_date = request.args.get("start_date")
        end_date = request.args.get("end_date")
        data = get_cashflow(start_date=start_date, end_date=end_date)
        return jsonify(data)
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@finance_bp.route("/api/finance/recurring")
def recurring():
    try:
        *_, get_recurring = _get_service()
        data = get_recurring()
        return jsonify(data)
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@finance_bp.route("/api/finance/net-worth-history")
def net_worth_history():
    try:
        from backend.services.monarch_service import get_net_worth_history
        months = request.args.get("months", 12, type=int)
        data = get_net_worth_history(months=months)
        return jsonify(data)
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@finance_bp.route("/api/finance/spending-breakdown")
def spending_breakdown():
    try:
        from backend.services.monarch_service import get_spending_breakdown
        months = request.args.get("months", 2, type=int)
        data = get_spending_breakdown(months=months)
        return jsonify(data)
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@finance_bp.route("/api/finance/cashflow-trend")
def cashflow_trend():
    try:
        from backend.services.monarch_service import get_cashflow_trend
        months = request.args.get("months", 6, type=int)
        data = get_cashflow_trend(months=months)
        return jsonify(data)
    except Exception as e:
        return jsonify({"error": str(e)}), 500
